Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/ShapeTool.kt

```diff
@@ -0,0 +1,175 @@
+private const val BUNDLE_OUTLINE_WIDTH = "OUTLINE_WIDTH"
+
+class ShapeTool(
+    shapeToolOptionsView: ShapeToolOptionsView,
+    contextCallback: ContextCallback,
+    toolOptionsViewController: ToolOptionsVisibilityController,
+    toolPaint: ToolPaint,
+    workspace: Workspace,
+    commandManager: CommandManager
+) : BaseToolWithRectangleShape(
+    contextCallback, toolOptionsViewController, toolPaint, workspace, commandManager
+) {
+    private val shapeToolOptionsView: ShapeToolOptionsView
+    private val shapePreviewPaint = Paint()
+    private val shapePreviewRect = RectF()
+    private val drawableFactory = DrawableFactory()
+    private var baseShape = DrawableShape.RECTANGLE
+    private var shapeDrawType = DrawableStyle.FILL
+    private var shapeDrawable = drawableFactory.createDrawable(baseShape)
+    private var shapeOutlineWidth = DEFAULT_OUTLINE_WIDTH
+    val shapeBitmapPaint = Paint()
+
+    override val toolType: ToolType
+        get() = ToolType.SHAPE
+
+    init {
+        rotationEnabled = true
+        this.shapeToolOptionsView = shapeToolOptionsView
+        this.shapeToolOptionsView.setCallback(
+            object : ShapeToolOptionsView.Callback {
+                override fun setToolType(shape: DrawableShape) {
+                    setBaseShape(shape)
+                    workspace.invalidate()
+                }
+
+                override fun setDrawType(drawType: DrawableStyle) {
+                    shapeDrawType = drawType
+                    workspace.invalidate()
+                }
+
+                override fun setOutlineWidth(outlineWidth: Int) {
+                    shapeOutlineWidth = outlineWidth
+                    workspace.invalidate()
+                }
+            })
+        toolOptionsViewController.showDelayed()
+    }
+
+    fun getBaseShape(): DrawableShape = baseShape
+
+    fun setBaseShape(shape: DrawableShape) {
+        baseShape = shape
+        shapeDrawable = drawableFactory.createDrawable(shape)
+    }
+
+    override fun changePaintColor(color: Int) {
+        super.changePaintColor(color)
+        workspace.invalidate()
+    }
+
+    override fun onSaveInstanceState(bundle: Bundle?) {
+        super.onSaveInstanceState(bundle)
+        bundle?.apply {
+            putSerializable(BUNDLE_BASE_SHAPE, baseShape)
+            putSerializable(BUNDLE_SHAPE_DRAW_TYPE, shapeDrawType)
+            putInt(BUNDLE_OUTLINE_WIDTH, shapeOutlineWidth)
+        }
+    }
+
+    override fun onRestoreInstanceState(bundle: Bundle?) {
+        super.onRestoreInstanceState(bundle)
+        val newBaseShape = bundle?.getSerializable(BUNDLE_BASE_SHAPE) as DrawableShape?
+        val newShapeDrawType = bundle?.getSerializable(BUNDLE_SHAPE_DRAW_TYPE) as DrawableStyle?
+        val newShapeOutlineWidth = bundle?.getInt(BUNDLE_OUTLINE_WIDTH)
+        if (newBaseShape != null && newShapeDrawType != null && newShapeOutlineWidth != null && (baseShape != newBaseShape || shapeDrawType != newShapeDrawType)) {
```

Comment:
Please, use `?: return` instead of ` != null`.

Reply:
done :)